document-generation/generate-document.ts

```diff
@@ -6,64 +6,68 @@ import inputOutputLogger from "@middy/input-output-logger";
 import errorLogger from "@middy/error-logger";
 import { logger } from "../utils/logging";
 import { tracer } from "../utils/tracing";
-import {
-  ApiBase64SuccessResponse,
-  OtherErrorResponse,
-  SuccessStatusCode,
-  ValidationErrorResponse,
-} from "../utils/response";
-import { INTERNAL_SERVER_ERROR, NOT_IMPLEMENTED } from "../utils/errors";
+import { ApiBase64SuccessResponse, SuccessStatusCode, ValidationErrorResponse } from "../utils/response";
+import { INTERNAL_SERVER_ERROR } from "../utils/errors";
 import { LoggingContextMiddleware } from "../utils/middleware/logging-context-middleware";
 import { errorHandlingMiddleware } from "../utils/middleware/error-handling-middleware";
 import JSONErrorHandlerMiddleware from "middy-middleware-json-error-handler";
 import validator from "@middy/validator";
 import xssSanitizer from "../utils/middleware/xss-sanitizer";
 import { wrapSchema } from "../utils/middleware/schema-wrapper";
 import { generateDocument } from "./chromium";
+import { generateIGCEDocument } from "./igce-document";
 import {
   generateDocumentSchema,
   RequestEvent,
   GenerateDocumentRequest,
   DocumentType,
+  TemplatePaths,
+  IndependentGovernmentCostEstimate,
 } from "../models/document-generation";
 import * as fs from "fs";
 import handlebars from "handlebars";
 import juice from "juice";
-import {
-  formatDuration,
-  formatGroupAndClassification,
-  counter,
-  countSections,
-  formatAwardType,
-} from "./handlebarUtils/helpers";
+import { formatDuration, formatGroupAndClassification, counter, countSections, formatAwardType } from "./utils/helpers";
 
 async function baseHandler(event: RequestEvent<GenerateDocumentRequest>): Promise<ApiBase64SuccessResponse> {
   const { documentType } = event.body;
+  logger.info("Generating document", { documentType });
+  const documentTemplatePaths: TemplatePaths = {
+    [DocumentType.DESCRIPTION_OF_WORK]: {
+      html: "/opt/dow-template.html",
+      css: "/opt/dow-style.css",
+    },
+    [DocumentType.INDEPENDENT_GOVERNMENT_COST_ESTIMATE]: {
+      excel: "/opt/igce-template.xlsx",
+    },
+  };
```

Comment:
Oh this would potentially be a really bad situation. Lambda runtimes can be reused and `/opt/` won't be emptied between. I don't _think_ we'd see concurrent executions in the same runtime. But if we did, that'd be even more catastrophic. And I _really_ don't want to end up in situations where we rewrite to the same file(s). Honestly, I'd guess we were making this mistake before but it's far more obvious now.

So I think that what we'd probably want to use is invoke [`fs.mkdtemp`](https://nodejs.org/docs/latest-v16.x/api/fs.html#fsmkdtempsyncprefix-options). Within that directory, we can use static filenames. But we've gotta use a different temp directory each time. And while I am being pedantic about this, we should use [`path.join`](https://nodejs.org/docs/latest-v16.x/api/path.html#pathjoinpaths) too.

```suggestion
  const baseDirectory = fs.mkdtemp("doc-gen");
  const documentTemplatePaths: TemplatePaths = {
    [DocumentType.DESCRIPTION_OF_WORK]: {
      html: path.join(baseDirectory, "dow", "template.html"),
      css: path.join(baseDirectory, "dow", "style.css"),
    },
    [DocumentType.INDEPENDENT_GOVERNMENT_COST_ESTIMATE]: {
      excel: path.join(baseDirectory, "igce", "template.xlsx"),
    },
  };
```

Reply:
Oh yeah that's an excellent point, those would be extremely bad 😵. It's easy to overlook the reuse of runtimes in the lambda, it didn't even cross my mind.

@kylelaker-ccpo so to be clear on your suggestion make a temp folder, copy the template files to that temp folder, and then have the lambda use a different temp folder for each execution?

Reply:
Ohh wait this is for the _inputs_?! That's my bad

Reply:
No I don't care about the input templates. Sorry. I shouldn't review when I'm tired.

Reply:
Yeah, the templates are the files being read in by the generating function as a base and then worked on inside of those functions. Only buffer are output and sent as the response, nothing should be saved as an output. Nothing is written to the original template file I believe.

It's all good, I appreciate the feedback. And now I'll be considering the runtime more.